File: src/class_conjugate_gradient.f90
!=====================================================================!
! The conjugate-gradient linear solver supplies the sweep (`iterate`)
! - one preconditioned cg pass driving the correction equation
! A dx = r - and inherits the residual-minimization iteration from
! linear_solver. Its optional preconditioner fills the inherited
! pre_conditioner slot (applied inside every cg iteration).
!
!    r --> z --> p --> A p --> alpha --> dx, res
!          ^                              |
!          '<------------ beta ----------'
!
! Matvecs and dots, nothing else. The newton/bdf linearized path that
! once lived here as a march override now lives on the system
! (linearize / clear_linearization): a frozen linear system answers
! the same two questions every system answers, so this solver cannot
! tell it apart from any other - and needs no override.
!=====================================================================!

module class_conjugate_gradient

  use iso_fortran_env         , only : dp => REAL64
  use interface_linear_solver , only : linear_solver, preconditioner
  use interface_assembler     , only : assembler

  implicit none

  ! Expose only the linear solver datatype.
  private
  public :: conjugate_gradient

  !===================================================================!
  ! The linear solver datatype.
  !===================================================================!

  type, extends(linear_solver) :: conjugate_gradient

   contains

     ! The sweep consumed by the inherited outer iteration.
     procedure :: iterate

  end type conjugate_gradient

  !===================================================================!
  ! The interface for multiple constructors.
  !===================================================================!

  interface conjugate_gradient
     module procedure construct
  end interface conjugate_gradient

contains

  !===================================================================!
  ! The constructor for the linear solver.
  !===================================================================!

  pure type(conjugate_gradient) function construct(max_it, &
       & max_tol, print_level, precond) result (this)

    type(integer)        , intent(in)           :: max_it
    type(real(dp))       , intent(in)           :: max_tol
    type(integer)        , intent(in)           :: print_level
    class(preconditioner), intent(in), optional :: precond

    this % max_it      = max_it
    this % max_tol     = max_tol
    this % print_level = print_level
    this % res_file    = 'cg.res'

    ! The optional preconditioner fills the pre-operator slot (plain CG without one).
    if (present(precond)) allocate(this % pre_conditioner, source = precond)

  end function construct

  !===================================================================!
  ! The sweep: one (preconditioned) conjugate-gradient pass driving the
  ! correction equation A dx = r from dx = 0.
  !===================================================================!

  impure subroutine iterate(this, system, r, dx, iter)

    class(conjugate_gradient) , intent(inout) :: this
    class(assembler)          , intent(in)    :: system
    real(dp)                  , intent(in)    :: r(:)
    real(dp)                  , intent(out)   :: dx(:)
    integer                   , intent(out)   :: iter

    !-----------------------------------------------------------------!
    ! Local data. The vector z holds the preconditioned residual
    ! M^-1 res (z = res when no pre-operator is attached, which gives
    ! plain CG).
    !-----------------------------------------------------------------!

    real(dp), allocatable :: p(:), res(:), w(:), z(:)
    real(dp)              :: alpha, beta
    real(dp)              :: bnorm, rnorm
    real(dp)              :: tol
    real(dp)              :: rho(2)

    iter = 1
    dx   = 0.0_dp

    allocate(p, res, w, z, mold = dx)

    !-----------------------------------------------------------------!
    ! The residual handed in is the right-hand side of the correction
    ! equation; at dx = 0 it is also the whole residual.
    !-----------------------------------------------------------------!

    res   = r
    bnorm = sqrt(system % inner_product(res, res))

    ! The homogeneous case leaves nothing to do.
    if (bnorm .le. this % max_tol) then
       iter = 0
       return
    end if

    rnorm = bnorm
    tol   = 1.0_dp

    !-----------------------------------------------------------------!
    ! The preconditioned residual is z = M^-1 res and rho = <res, z>
    ! (PCG). When unpreconditioned z = res, so rho = <res, res> and
    ! this is plain CG.
    !-----------------------------------------------------------------!

    call this % apply_pre_conditioner(res, z)
    rho(2) = system % inner_product(res, z)

    ! Apply the iterative scheme until the tolerance is achieved.
    do while ((tol .gt. this % max_tol) .and. (iter .lt. this % max_it))

       ! Step (a): form the descent direction (on the preconditioned residual).
       if ( iter .eq. 1) then
          p = z
       else
          beta = rho(2)/rho(1)
          p = z + beta*p
       end if

       ! Step (b): apply the operator action on the direction.
       call system % get_jacobian_residual_product(w, p)

       ! Step (c): compute the step size.
       alpha = rho(2)/system % inner_product(p, w)

       ! Step (d): update the correction.
       dx = dx + alpha*p

       ! Step (e): form the new residual (the true residual drives the tolerance).
       res = res - alpha*w

       ! Step (f): update values before the next iteration.
       rnorm = sqrt(system % inner_product(res, res))
       tol   = rnorm/bnorm

       if (this % print_level .gt. 1) then
          write(*,*) iter, tol, rnorm, rho
       end if

       iter = iter + 1

       call this % apply_pre_conditioner(res, z)
       rho(1) = rho(2)
       rho(2) = system % inner_product(res, z)

    end do

    !-----------------------------------------------------------------!
    ! Report the CG steps taken (iter starts at 1, so iter - 1 equals
    ! the steps); the outer iteration accumulates this on the object.
    !-----------------------------------------------------------------!

    iter = iter - 1

    deallocate(res, p, w, z)

  end subroutine iterate

end module class_conjugate_gradient

File: ufvm/class_conjugate_gradient.py
#=====================================================================#
# Conjugate Gradient linear solver class that uses the functionalities
# of assembler class in the iterative solution process.
#=====================================================================#

import numpy as np

from .interface_linear_solver import LinearSolver


class ConjugateGradient(LinearSolver):

    #===================================================================!
    # Constructor for linear solver
    #
    # Fortran: interface conjugate_gradient => construct(FVAssembler,
    #          max_it, max_tol, print_level)
    #===================================================================!
    def __init__(self, FVAssembler, max_it, max_tol, print_level):
        self.FVAssembler = FVAssembler
        self.max_it = max_it
        self.max_tol = max_tol
        self.print_level = print_level

    #===================================================================!
    # Iterative linear solution using conjugate gradient method
    #===================================================================!
    def solve(self):
        # Initial guess vector for the subspace is "b"
        x = np.zeros(self.FVAssembler.num_state_vars, dtype=np.float64)
        self.FVAssembler.get_source(x)
        if np.linalg.norm(x) < np.finfo(np.float64).eps:
            print('zero rhs? stopping')
            raise SystemExit

        xold = np.zeros(self.FVAssembler.num_state_vars, dtype=np.float64)
        ss = np.zeros(self.FVAssembler.num_state_vars, dtype=np.float64)

        res_file = None
        if self.print_level == -1:
            res_file = open('cg.res', 'w')
            res_file.write(" iteration  residual\n")

        update_norm = np.finfo(np.float64).max
        iter = 1
        while update_norm > self.max_tol:

            xold[:] = x

            # Inner iterations with CG
            if iter == 1:
                ss[:] = 0.0
                num_inner_iters = self.iterate(x, ss)
            else:
                self.FVAssembler.get_skew_source(ss, x)
                num_inner_iters = self.iterate(x, ss)

            update_norm = np.linalg.norm(x - xold)
            if self.print_level == -1:
                res_file.write("%d %s\n" % (iter, repr(float(update_norm))))
            if self.print_level > 0:
                print("outer iter", iter, "num_inner_iters", num_inner_iters,
                      "update norm", update_norm)
            iter = iter + 1

        if self.print_level == -1:
            res_file.close()

        return x

    #===================================================================!
    # Iterative linear solution using conjugate gradient method
    #===================================================================!
    def iterate(self, x, ss):
        # Start the iteration counter
        iter = 1

        # Memory allocations
        b = np.zeros_like(x)
        p = np.zeros_like(x)
        r = np.zeros_like(x)
        w = np.zeros_like(x)
        Ax = np.zeros_like(x)
        tmp = np.zeros_like(x)

        rho = np.zeros(2, dtype=np.float64)

        # Norm of the right hand side
        self.FVAssembler.get_source(tmp)
        # Add the additional right hand side supplied
        tmp = tmp + ss
        if self.FVAssembler.symmetry is False:
            self.FVAssembler.get_transpose_jacobian_vector_product(b, tmp)
        else:
            b = tmp.copy()
        bnorm = np.linalg.norm(b)

        # Homogeneous case
        if bnorm <= self.max_tol:
            x[:] = 0.0
            return iter

        # Norm of the initial residual
        self.FVAssembler.get_jacobian_vector_product(tmp, x)
        if self.FVAssembler.symmetry is False:
            self.FVAssembler.get_transpose_jacobian_vector_product(Ax, tmp)
        else:
            Ax = tmp.copy()
        r = b - Ax
        rnorm = np.linalg.norm(r)
        tol = rnorm / bnorm
        rho[2 - 1] = rnorm * rnorm

        # Apply Iterative scheme until tolerance is achieved
        while (tol > self.max_tol) and (iter < self.max_it):

            # step (a) compute the descent direction
            if iter == 1:
                # steepest descent direction p
                p = r.copy()
            else:
                # take a conjugate direction
                beta = rho[2 - 1] / rho[1 - 1]
                p = r + beta * p

            # step (b) compute the solution update
            self.FVAssembler.get_jacobian_vector_product(tmp, p)
            if self.FVAssembler.symmetry is False:
                self.FVAssembler.get_transpose_jacobian_vector_product(w, tmp)
            else:
                w = tmp.copy()

            # step (c) compute the step size for update
            alpha = rho[2 - 1] / np.dot(p, w)

            # step (d) Add dx to the old solution
            x[:] = x + alpha * p

            # step (e) compute the new residual
            r = r - alpha * w

            # step(f) update values before next iteration
            rnorm = np.linalg.norm(r)
            tol = rnorm / bnorm

            if self.print_level > 1:
                print(iter, tol, rnorm, rho)

            iter = iter + 1

            rho[1 - 1] = rho[2 - 1]
            rho[2 - 1] = rnorm * rnorm

        return iter
